File: app/src/main/java/com/petal/browser/ui/components/ElasticOverscrollHelper.java
package com.petal.browser.ui.components;

/**
 * ElasticOverscrollHelper provides spring resistance physics calculations for overscroll rubber-banding.
 */
public class ElasticOverscrollHelper {

    /**
     * Calculates rubber-band offset with logarithmic resistance decay.
     */
    public static float calculateRubberBandOffset(float delta, float maxOffset) {
        if (delta == 0) return 0f;
        float sign = delta > 0 ? 1f : -1f;
        float absDelta = Math.abs(delta);
        float rubberBand = (1f - (1f / ((absDelta * 0.55f / maxOffset) + 1f))) * maxOffset;
        return sign * rubberBand;
    }
}

File: app/src/main/java/com/petal/browser/ui/components/ElasticOverscrollHelper.kt
package com.petal.browser.ui.components

object ElasticOverscrollHelper {

    /**
     * Calculates rubber-band offset with logarithmic resistance decay.
     */
    @JvmStatic
    fun calculateRubberBandOffset(delta: Float, maxOffset: Float): Float {
        if (delta == 0f) return 0f
        val sign = if (delta > 0f) 1f else -1f
        val absDelta = Math.abs(delta)
        val rubberBand = (1f - 1f / (absDelta * 0.55f / maxOffset + 1f)) * maxOffset
        return sign * rubberBand
    }
}
